### PR TITLE
API: ArrowExtensionArray._cmp_method to return pyarrow.bool_ type

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file if fixing a bug or adding a new feature.

Per https://github.com/pandas-dev/pandas/issues/51462#issuecomment-1442527928, updating `ArrowExtensionArray._cmp_method` to return `boolean[pyarrow]` dtype instead of `boolean`.

### CODE REVIEW

File: asv_bench/asv.conf.json [hunk 1]
```diff
@@ -126,6 +126,6 @@
     },
     "build_command":
     ["python -m pip install versioneer[toml]",
-     "python setup.py build -j4",
+     "python setup.py build -j1",
```
Comment (phofl) [inline asv_bench/asv.conf.json:129]: Looks like this is failing asvs?

File: pandas/core/arrays/arrow/array.py [hunk 2]
```diff
@@ -81,10 +81,10 @@
...
     ARROW_LOGICAL_FUNCS = {
-        "and": pc.and_kleene,
-        "rand": lambda x, y: pc.and_kleene(y, x),
-        "or": pc.or_kleene,
-        "ror": lambda x, y: pc.or_kleene(y, x),
+        "and_": pc.and_kleene,
+        "rand_": lambda x, y: pc.and_kleene(y, x),
+        "or_": pc.or_kleene,
+        "ror_": lambda x, y: pc.or_kleene(y, x),
         "xor": pc.xor,
```
Comment (lukemanley) [inline pandas/core/arrays/arrow/array.py:88]: I think(?) the first four here should have underscores. These were raising in an ASV.

File: asv_bench/asv.conf.json (hunk 1 above)
Comment (lukemanley) [reply asv_bench/asv.conf.json]: I've reverted this. The asv job was failing before I tried this change. This was just an attempt to find the issue.

File: pandas/core/arrays/arrow/array.py (hunk 2 above)
Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py:88]: for example, this was raising:

```
import pandas as pd

s1 = pd.Series([True], dtype="boolean[pyarrow]")
s2 = pd.Series([False], dtype="boolean[pyarrow]")

# raises: KeyError: 'or_'
s1 | s2  
```

Comment (phofl) [reply pandas/core/arrays/arrow/array.py:88]: Good find! Could you move into a separate pr and add tests?

Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py:88]: sure, will do

Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py:88]: opened #51688 for this

**Review by phofl — COMMENTED** [review]
the min versions failure looks legit?

**Comment by lukemanley** [comment]
> the min versions failure looks legit?

yep. just pushed a fix. I think the failure is related to the `deepcopy` you added in https://github.com/pandas-dev/pandas/pull/50984

It looks like that test combined with the deepcopy was hitting this bug in older versions of numpy: https://github.com/numpy/numpy/issues/21883

I think this will work now. What is the reason to use deepcopy instead of `ndarray.copy` there? It can be very slow for an object dtype array:

```
import numpy as np
from copy import deepcopy

arr = np.arange(1_000_000, dtype=object)

%timeit arr.copy()
%timeit deepcopy(arr)

# 9.83 ms ± 757 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# 298 ms ± 4.82 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)    
```

**Comment by phofl** [comment]
@mroeschke argued back then that this would cover eas without a copy method, but if this is slow, then I'd say we can switch to copy

**Comment by mroeschke** [comment]
Fine to switch to `copy()`

File: pandas/core/arrays/arrow/array.py [hunk 3]
```diff
@@ -581,6 +572,122 @@ def isna(self) -> npt.NDArray[np.bool_]:
 
         return self._data.is_null().to_numpy()
 
+    def any(self, *, skipna: bool = True, **kwargs):
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:575]: I guess technically `ArrowExtensionArray` should subclass `ExtensionArraySupportsAnyAll` now instead?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lukemanley

**Comment by phofl** [comment]
@mroeschke thoughts on back porting? This would fix #51980

**Comment by mroeschke** [comment]
Sure, +1 to backport

**Pull request merged**